docs/source/trainer/using_the_trainer.rst

```diff
@@ -0,0 +1,484 @@
+
+
+Callbacks
+~~~~~~~~~
+
+You can insert arbitrary callbacks to be run at various points during
+the training loop. The Composer library provides several useful
+callbacks for things such as monitoring throughput and memory usage
+during training, but you can also implement your own.
+
+.. code:: python
+
+   from composer import Trainer
+   from composer.callbacks import SpeedMonitor
+
+   # include a callback for tracking throughput/step during training
+   trainer = Trainer(model=model,
+                     train_dataloader=train_dataloader,
+                     eval_dataloader=eval_dataloader,
+                     max_duration='160ep',
+                     device='gpu',
+                     callbacks=[SpeedMonitor(window_size=100)])
+
+.. seealso::
+
+    The :doc:`Callbacks<trainer/callbacks>` guide and :mod:`composer.callbacks`.
+
+
+Numerics
+~~~~~~~~
+
+Using `mixed precision <https://arxiv.org/abs/1710.03740>`__ can speed up your training loop and only requires
+setting the ``precision`` parameter in the trainer. Note that ``amp``
+only works if training on a GPU.
+
+.. code:: python
+
+   from composer import Trainer
+
+   # use mixed precision during training
+   trainer = Trainer(model=model,
+                     train_dataloader=train_dataloader,
+                     eval_dataloader=eval_dataloader,
+                     max_duration='160ep',
+                     device='gpu',
+                     precision='amp')
+
+Checkpointing
+~~~~~~~~~~~~~
+
+The Composer trainer makes it easy to (1) save checkpoints at various
+points during training and (2) load them back to resume training later.
+
+.. code:: python
+
+   from composer import Trainer
+
+   ### Saving checkpoints
+   trainer = Trainer(model=model,
+                     train_dataloader=train_dataloader,
+                     eval_dataloader=eval_dataloader,
+                     max_duration='160ep',
+                     device='gpu',
+                     # Checkpointing params
+                     save_folder: 'checkpoints',
+                     save_interval: '1ep')
+
+   # will save checkpoints to the 'checkpoints' folder every epoch
+   trainer.fit()
+
+.. code:: python
+
+   from composer import Trainer
+
+   ### Loading checkpoints
+   trainer = Trainer(model=model,
+                     train_dataloader=train_dataloader,
+                     eval_dataloader=eval_dataloader,
+                     max_duration='160ep',
+                     device='gpu',
+                     # Checkpointing params
+                     load_path: 'path/to/checkpoint/mosaic_states.pt')
+
+   # will load the trainer state (including model weights) from the
+   # load_path before resuming training
+   trainer.fit()
+
+.. seealso::
+
+    The :doc:`Checkpointing<trainer/checkpointing>` guide.
+
+
+This was just a quick tour of all the features within our trainer. Please see the other
+guides and notebooks for more information.
+
+Annotated Trainer Loop
```

Comment:
The training loop code should be annotated via docstrings. Can we refactor the training loop code for readability? Ideally each function in the training loop should be as long as the pseudocode here so it is readable.

Reply:
This is out of scope for this PR